### ISSUE TITLE
batch insert/upsert versus batch AQL insert/upsert: AQL is 100 to 2100 times slower?

### ISSUE THREAD

**Issue by coldcoff** [arangodb/arangodb#2523]
Hi!

I'm currently trying to write data to ArangoDB to work with it in a later step.

I figured out that using AQL (insert/replace/upsert) in batches is **significantly** slower than importing it with the collection methods (insert_many, replace_many/import_bulk)

1000 inserts/replace/upsert via collection methods take roughly 0.03 seconds (number of shards is irrelevant)
1000 inserts/replace/upsert via AQL take roughly 6 seconds, when sharding (6 seconds * number of shards)

That means: Having 3 machines and the collection configured to utilize 9 shards the difference is 0.03 seconds to 1:05 minutes for 1000 insert/replace/update. This is then 2100 times slower?!

```...
committing 1000 inserts ...
commit done, duration = 0:00:00.017035
committing 1000 AQL jobs ...
commit done, duration = 0:01:04.284003
...
```
As server I'm using ArangoDB 3.1.20 on CentOS, sharded, 3 machines (shard_count set to either 1 or 9 in the example) and ArangoDB 3.1.21 on Ubuntu 16.04, 1 machine, no shards.

For the script I'm using the python-arango interface but this doesn't seem to matter, profiling the script shows that the long time is waiting for the REST answers.

A script showing that behaviour is attached for reference: [slowaql.zip](https://github.com/arangodb/arangodb/files/1054342/slowaql.zip)

Thanks

coldcoff

Here's the Form:

## my environment running ArangoDB
I'm using the latest ArangoDB of the respective release series:
- [ ] 2.8
- [ ] 3.0
- [x] 3.1
- [ ] 3.2 pre-releases: ___
- [ ] self-compiled devel branch

On this operating system:
- [ ] DCOS on
  - [ ] AWS
  - [ ] Azure
  - [ ] own infrastructure
- [x] Linux 
  - [ ] Debian .deb
  - [x] Ubuntu .deb
  - [ ] SUSE   .rpm
  - [x] RedHat .rpm
  - [ ] Fedora .rpm
  - [ ] Gentoo
  - [ ] docker - official docker library
  - [ ] other:
- [ ] Windows, version: 
- [ ] MacOS, version:

### this is an AQL-related issue:
[ ] I'm using graph features

I'm issuing AQL via:
- [ ] web interface with this browser:     running on this OS:
- [ ] arangosh
- [x] this Driver: python-arango, version 3.7.0

I've run `db._explain("<my aql query>")` and it didn't shed more light on this.
The AQL query in question is:
**see attached script 'slowaql.py'**
[slowaql.zip](https://github.com/arangodb/arangodb/files/1054342/slowaql.zip)

The issue can be reproduced using this dataset:
**see attached script 'slowaql.py'**
[slowaql.zip](https://github.com/arangodb/arangodb/files/1054342/slowaql.zip)

Please provide a way to create the dataset to run the above query on; either by a gist with an arangodump, or `db.collection.save({my: "values"}) statements. If it can be reproduced with one of the ArangoDB example datasets, it's a plus.

### Foxx
-

### this is a web interface-related issue: NO
I'm using the web interface with this browser:     running on this OS:
- [ ] authentication is enabled?
- [ ] using the cluster?
- [ ] _system database?

These are the steps to reproduce:
1) open the browser on http://127.0.0.1:8529
2) log in as ...
3) use database [ ] `_system` [ ] other: 
4) click to ...
...

The following problem occurs:
```
committing 1000 inserts ...
commit done, duration = 0:00:00.070130
committing 1000 inserts ...
commit done, duration = 0:00:00.036639
committing 1000 inserts ...
commit done, duration = 0:00:00.029238
committing 1000 inserts ...
commit done, duration = 0:00:00.030998
committing 1000 inserts ...
commit done, duration = 0:00:00.030512
committing 1000 inserts ...
commit done, duration = 0:00:00.034857
committing 1000 inserts ...
commit done, duration = 0:00:00.038439
committing 1000 inserts ...
commit done, duration = 0:00:00.031574
committing 1000 inserts ...
commit done, duration = 0:00:00.029863
committing 1000 inserts ...
commit done, duration = 0:00:00.028538
committing 1000 AQL jobs ...
commit done, duration = 0:00:06.596808
committing 1000 AQL jobs ...
commit done, duration = 0:00:05.914047
committing 1000 AQL jobs ...
commit done, duration = 0:00:06.357795
committing 1000 AQL jobs ...
commit done, duration = 0:00:06.464578
committing 1000 AQL jobs ...
commit done, duration = 0:00:06.426082
committing 1000 AQL jobs ...
commit done, duration = 0:00:06.428333
committing 1000 AQL jobs ...
commit done, duration = 0:00:06.744315
committing 1000 AQL jobs ...
commit done, duration = 0:00:06.693037
committing 1000 AQL jobs ...
commit done, duration = 0:00:06.766508
committing 1000 AQL jobs ...
commit done, duration = 0:00:06.340712
```
(1 shard)

Instead I would be expecting: 
**roughly the same execution times** 

Note that increasing the shard_count on the collection will dramatically increase the needed AQL time but leave the collection method time nearly the same.

### this is an installation-related issue: NO
Describe which steps you carried out, what you expected to happen and what actually happened.

**Comment by coldcoff** [comment]
Any news on that? At least a classification? Is such a massive slowdown intended or a bug?

**Comment by dothebart** [comment]
Its rather a question of perspective. 

Its moreof that the import facility is massively faster than what you've used to progamatically insert data. 

This can only be achieved by the import facility making certain assumptions which the AQL version can't take.

Thus we advice to use the import facility for bulk insert over using AQL queries like you did. 
You can use [arangoimp](https://docs.arangodb.com/latest/Manual/Administration/Arangoimp.html) for this, or if you want to do something programatically [use the API arangoimp also uses](https://docs.arangodb.com/latest/HTTP/BulkImports/) under the hood.

We'll give you a more detailed answer once the next ArangoDB 3.2 Beta release is done.

**Comment by coldcoff** [comment]
Thanks @dothebart.

For now I can work with import_bulk(), which is fast, so it's OK to wait for you to finish ArangoDB 3.2 beta first.

To document / illustrate the slowness see the following numbers, extracted differently from the very same script above:

a) ArangoDB 3.1.22, cluster setup (3 machines). collection with shard_count = 1 (!)

> Elapsed 23.68531ms/1000 inserts
> Elapsed 5850.9199ms/1000 AQLs

b) ArangoDB 3.1.22, *no* cluster setup (just one of the 3 machines as single DB). collection with shard_count = 1

> Elapsed 16.20139ms/1000 inserts
> Elapsed 364.672ms/1000 AQLs

c) ArangoDB 3.1.22, cluster setup (3 machines). collection with shard_count = 3

> Elapsed 22.74879ms/1000 inserts
> Elapsed 25556.0558ms/1000 AQLs

**Comment by francoisprunier** [comment]
Hi,

I'm seeing some very weird slowdowns with AQL as well.
Version 3.1.22

My workload:
- multiple HTTP batch requests /_api/batch
- each batch consists of a mix of 5000
  - POST /_api/document/ to a vertex collection
  - POST /_api/document/ to an edge collection
  - AQL INSERT requests to add more edges to the edge collection
- after all calls have returned: 130 000 vertices created, 561 000 edges created

The same workload executed in different conditions but *always one shard, no replication*:
- single server setup: 	Data imported in 23 345ms
- cluster setup with one agency, one coordinator, one primary server: 	Data imported in 624 020ms
- cluster setup with one agency, one coordinator, one primary server BUT I removed the AQL calls: 	Data imported in 32 943ms

Clearly there's something wrong going on with AQL and clustering. From my point of view, it makes the clustering not scale horizontally at all.

I'm happy to provide more information if needed.

**Comment by francoisprunier** [comment]
You can replicate it with arangobench

Cluster mode (one agency, one coordinator, one primary)
```
Total number of operations: 10000, runs: 1, keep alive: yes, async: no, batch size: 1000, replication factor: 1, number of shards: 1, wait for sync: false, concurrency level (threads): 10
Test case: aqlinsert, complexity: 20, database: '_system', collection: 'ArangoBenchmark'
Total request/response duration (sum of all threads): 271.256008 s
Request/response duration (per thread): 27.125601 s
Time needed per operation: 0.002781 s
Time needed per operation per thread: 0.027813 s
Operations per second rate: 359.540881
Elapsed time since start: 27.813249 s
```

Single server:
```
Total number of operations: 10000, runs: 1, keep alive: yes, async: no, batch size: 1000, replication factor: 1, number of shards: 1, wait for sync: false, concurrency level (threads): 10
Test case: aqlinsert, complexity: 20, database: '_system', collection: 'ArangoBenchmark'
Total request/response duration (sum of all threads): 6.498392 s
Request/response duration (per thread): 0.649839 s
Time needed per operation: 0.000074 s
Time needed per operation per thread: 0.000736 s
Operations per second rate: 13579.094760
Elapsed time since start: 0.736426 s
```

That's a 41x slow down.

**Comment by francoisprunier** [comment]
@dothebart could you give us your opinion on this please ?

**Comment by dsonet** [comment]
@dothebart Is there any updates on this issue?

**Comment by dothebart** [comment]
you would probably better hilight @kvahed or @neunhoef on this case.

**Comment by nkanand4** [comment]
I have a requirement of bulk upload/insert of documents and need to create/update edges accordingly. Here we need to make sure that during this batch insert of documents and edges are not duplicates. So, the result of this operation should be idempotent and need me to use upsert in AQL. After around 6000 records, things start getting slow to a point where it feels it is stuck. Any update on this issue?

**Comment by OmarAyo** [comment]
Hi all,

This ticket is being closed as inactive due to the date of the last activity on it.

If you believe this was done in error, please comment or create a new issue.
